File: clean_raw_eeg.m
clc;
clear all;
close all;

%% 1. Setup Paths (CHANGE THESE TWO LINES IF NEEDED)
% -> Path to EEGLAB
eegpath = 'C:\Users\dppcs\Downloads\eeglab_current\eeglab2026.0.0'; 
addpath(eegpath);
eeglab;

% -> Path to the main Neuroflow folder
rootDir = 'C:\Users\dppcs\OneDrive\Desktop\DPP_Neuroflow'; 

%% 2. Find all EEG folders and the LARGEST EDF files
fprintf('🔍 Scanning %s for subject folders...\n', rootDir);
eegFolders = dir(fullfile(rootDir, '**', '*_EEG'));
edfPaths = {};
layPaths = {};

for k = 1:length(eegFolders)
    folderPath = fullfile(eegFolders(k).folder, eegFolders(k).name);
    edfFiles = dir(fullfile(folderPath, '*.edf'));
    
    if isempty(edfFiles)
        continue;
    end
    
    % --- THE AUTOPILOT LOGIC: IF MULTIPLE FILES, PICK THE LARGEST ONE ---
    if length(edfFiles) > 1
        fprintf('⚠️ Found %d EDFs in %s. Selecting the largest one to avoid false starts...\n', length(edfFiles), eegFolders(k).name);
        [~, maxIdx] = max([edfFiles.bytes]); % Find the index of the biggest file
        edfFiles = edfFiles(maxIdx);         % Keep ONLY the biggest file
    end
    
    currentEdfPath = fullfile(edfFiles(1).folder, edfFiles(1).name);
    edfPaths{end+1} = currentEdfPath; %#ok<SAGROW>
    
    % Find the matching .lay file
    [~, baseName, ~] = fileparts(currentEdfPath);
    layFile = dir(fullfile(folderPath, [baseName, '.lay']));
    if ~isempty(layFile)
        layPaths{end+1} = fullfile(layFile(1).folder, layFile(1).name); %#ok<SAGROW>
    else
        layPaths{end+1} = ''; 
    end
end

fprintf('✅ Found %d valid (largest) EDF files to process.\n', length(edfPaths));

%% 3. Process Each File (The Loop)
for fileIdx = 1:length(edfPaths)
    currentEdf = edfPaths{fileIdx};
    currentLay = layPaths{fileIdx};
    [filepath, name, ~] = fileparts(currentEdf);
    
    fprintf('\n=========================================================\n');
    fprintf('⚙️ Processing File %d of %d: %s\n', fileIdx, length(edfPaths), name);
    
    % --- Read .lay file for info ---
    info = struct();
    info.sample_rate = 250; % Default fallback
    if ~isempty(currentLay) && isfile(currentLay)
        fid = fopen(currentLay, 'r');
        lines = textscan(fid, '%s', 'Delimiter', '\n');
        fclose(fid);
        lines = lines{1};
        for i = 1:length(lines)
            line = strtrim(lines{i});
            if contains(line, 'samplerate=') || contains(line, 'samplerate =')
                parts = strsplit(line, '=');
                info.sample_rate = str2double(strtrim(parts{2}));
            end
        end
    end
    
    % --- Load EDF Data ---
    try
        fprintf('   -> Loading EDF data...\n');
        data = edfread(currentEdf);
        
        % --- THE 8 CHANNELS FOR MAESTRO ---
        select_channels = {'FZ', 'CZ', 'C3', 'C4', 'O1', 'O2', 'F3_BLUE_', 'F4_RED_'}; 
        standard_names = {'Fz', 'Cz', 'C3', 'C4', 'O1', 'O2', 'F3', 'F4'};
        
        data_subset = data(:, select_channels);
        
        % Unpack the table into an array
        first_col = data_subset{:, 1}; 
        if iscell(first_col)
            total_samples = sum(cellfun(@numel, first_col));
        else
            total_samples = numel(first_col);
        end
        
        data_new = zeros(length(select_channels), total_samples);
        for i = 1:length(select_channels)
            col_data = data_subset{:, i};
            if iscell(col_data)
                data_new(i, :) = vertcat(col_data{:})';
            else
                data_new(i, :) = col_data(:)';
            end
        end
        
        % --- Import to EEGLAB ---
        fprintf('   -> Importing to EEGLAB...\n');
        EEG = pop_importdata('dataformat', 'array', 'data', data_new, 'srate', info.sample_rate);
        
        % Rename to standard names EXACTLY ONCE, then lookup coordinates
        for i = 1:length(standard_names)
            EEG.chanlocs(i).labels = standard_names{i}; 
        end
        EEG = pop_chanedit(EEG, 'lookup','standard-10-5-cap385.elp');
        
        % --- Filtering ---
        fprintf('   -> Filtering noise (0.5 - 35 Hz)...\n');
        EEG_filt = pop_eegfiltnew(EEG, 'locutoff',  0.5, 'hicutoff',  35, 'filtorder', 3300, 'plotfreqz', 0);
        
        % --- Resampling ---
        fprintf('   -> Resampling to 500 Hz for MAESTRO...\n');
        EEG_filt = pop_resample(EEG_filt, 500);
        
        % --- Automatic Artifact Removal (ICA) ---
        fprintf('   -> Running ICA (Artifact Removal). This will take a moment...\n');
        EEG_filt = pop_runica(EEG_filt, 'icatype', 'runica', 'interrupt','off');
        EEG_filt = iclabel(EEG_filt);
        
        threshold_signal = 0.1; 
        cls_score = EEG_filt.etc.ic_classification.ICLabel.classifications;
        bad_comp = [];
        for cmp=1:size(EEG_filt.icachansind,2)
            if cls_score(cmp,1) < threshold_signal
                bad_comp = [bad_comp, cmp];
            end
        end
        fprintf('   -> Removing %d noisy components...\n', length(bad_comp));
        EEG_ica = pop_subcomp(EEG_filt, bad_comp, 0);
        
        % --- Re-reference ---
        EEG_reref = pop_reref(EEG_ica, []);
        
        % --- SAVE THE CLEAN DATA ---
        saveFileName = sprintf('%s_cleaned_EEG_500Hz.csv', name);
        savePath = fullfile(filepath, saveFileName);
        
        % Transpose so it saves as [Time x 8 Channels]
        writematrix(EEG_reref.data', savePath);
        fprintf('   ✅ Saved pristine 500Hz EEG to: %s\n', savePath);
        
    catch ME
        fprintf('   ❌ Error processing %s: %s\n', currentEdf, ME.message);
    end
end
fprintf('\n🎉 All multi-subject EEG preprocessing complete!\n');

File: clean_raw_eeg.py
import os
import glob
import mne
from mne_icalabel import label_components
import pandas as pd
import numpy as np

# ==========================================
# 1. SETUP PATHS
# ==========================================
root_dir = '/home/stud1/Desktop/Swathi/COGBCI_data_codes/Neuroflow/raw_data'

# 🚨 THE UPGRADE: Automatically find EVERY subject folder in raw_data
# This ignores hidden files/folders and sorts them neatly
subjects = [d for d in os.listdir(root_dir) if os.path.isdir(os.path.join(root_dir, d)) and not d.startswith('.')]
subjects.sort()

print(f"🔍 Scanning {root_dir}...")
print(f"✅ Found {len(subjects)} subjects to process.\n")

# ==========================================
# 2. PROCESSING LOOP
# ==========================================
for subj in subjects:
    subject_path = os.path.join(root_dir, subj)
    
    # Find the EDF recursively
    edf_files = glob.glob(os.path.join(subject_path, '**', '*.edf'), recursive=True)
    if not edf_files:
        print(f"⚠️ No EDF found for Subject: {subj}")
        continue
        
    # Pick the largest file to avoid tiny calibration/aborted files
    current_edf = max(edf_files, key=os.path.getsize)
    print(f"\n=========================================================")
    print(f"⚙️ Processing Subject {subj}: {os.path.basename(current_edf)}")
    
    try:
        # Load Data
        print("   -> Loading EDF...")
        raw = mne.io.read_raw_edf(current_edf, preload=True, verbose=False)
        
        # SMART CHANNEL SELECTION (Handles MATLAB/Python naming differences)
        available_chans = raw.ch_names
        target_channels = {
            'Fz': ['FZ', 'Fz', 'fz'],
            'Cz': ['CZ', 'Cz', 'cz'],
            'C3': ['C3', 'c3'],
            'C4': ['C4', 'c4'],
            'O1': ['O1', 'o1'],
            'O2': ['O2', 'o2'],
            'F3': ['F3(BLUE)', 'F3_BLUE_', 'F3-BLUE', 'F3_BLUE', 'F3', 'f3'],
            'F4': ['F4(RED)', 'F4_RED_', 'F4-RED', 'F4_RED', 'F4', 'f4']
        }
        
        channel_mapping = {}
        missing_targets = []
        
        for std_name, variations in target_channels.items():
            match = next((var for var in variations if var in available_chans), None)
            if match:
                channel_mapping[match] = std_name
            else:
                missing_targets.append(std_name)
                
        if missing_targets:
            raise ValueError(f"Missing {missing_targets}. Available: {available_chans}")
            
        # Drop unwanted channels and rename the 8 we need
        raw.pick(list(channel_mapping.keys()))
        raw.rename_channels(channel_mapping)
        
        # Set standard 10-20 montage coordinates
        montage = mne.channels.make_standard_montage('standard_1020')
        raw.set_montage(montage)
        
        # Filter & Resample
        print("   -> Filtering (0.5 - 35 Hz) and Resampling to 500Hz...")
        raw.filter(l_freq=0.5, h_freq=35.0, verbose=False)
        raw.resample(500.0, verbose=False)
        
        # Run ICA and Auto-Reject Artifacts
        print("   -> Running ICA...")
        ica = mne.preprocessing.ICA(n_components=len(raw.ch_names), random_state=42, max_iter='auto')
        ica.fit(raw, verbose=False)
        
        ic_labels = label_components(raw, ica, method='iclabel')
        
        bad_comps = [idx for idx, label in enumerate(ic_labels['labels']) if label != 'brain']
        ica.exclude = bad_comps
        
        print(f"   -> Removed {len(bad_comps)} non-brain components.")
        
        clean_raw = ica.apply(raw.copy(), verbose=False)
        
        # Average Reference
        clean_raw.set_eeg_reference('average', projection=False, verbose=False)
        
        # Save to CSV [Time x 8 Channels]
        save_path = current_edf.replace('.edf', '_cleaned_EEG_500Hz.csv')
        data_array = clean_raw.get_data()
        
        pd.DataFrame(data_array.T).to_csv(save_path, index=False, header=False)
        print(f"   ✅ Successfully Saved: {save_path}")

    except Exception as e:
        print(f"   ❌ FAILED on Subject {subj}: {e}")

print("\n🎉 ALL SUBJECTS PROCESSED!")
